In [1]:
# Import nessesarry libraries
import pandas as pd
import numpy as np

### Read CSV file

Link dataset: https://www.kaggle.com/datasets/linhttht/car-price-vn/data?select=data.csv

In [4]:
# Read  file 
df = pd.read_csv('data_carVN_raw.csv',on_bad_lines='skip', encoding='UTF-8-SIG',na_values=['nan','-'])

In [5]:
df.head(3)

car_name    year            price  \
0          LandRover Range Rover HSE 3.0  2015.0   2 Tỷ 700 Triệu   
1   Mitsubishi Outlander 2.0 CVT Premium  2018.0        666 Triệu   
2                    Mazda 3 1.5L Luxury  2019.0        490 Triệu   

       assemble_place series  driven kms  num_of_door  num_of_seat  \
0           Nhập khẩu    SUV     65000.0          5.0          5.0   
1  Lắp ráp trong nước    SUV     50000.0          5.0          7.0   
2  Lắp ráp trong nước  Sedan     51000.0          4.0          5.0   

  engine_type transmission                                                url  
0        Xăng   Số tự động  https://bonbanh.com/xe-landrover-range_rover-h...  
1        Xăng   Số tự động  https://bonbanh.com/xe-mitsubishi-outlander-2....  
2        Xăng   Số tự động  https://bonbanh.com/xe-mazda-3-1.5l-luxury-201...

In [6]:
df.index

RangeIndex(start=0, stop=27484, step=1)

DataFrame starts from 0, ends before 27484, with a step size of 1

### Drop unnecessary column ('url' column)

In [9]:
df = df.drop(['url'],axis = 1)

In [10]:
df.head(3)

car_name    year            price  \
0          LandRover Range Rover HSE 3.0  2015.0   2 Tỷ 700 Triệu   
1   Mitsubishi Outlander 2.0 CVT Premium  2018.0        666 Triệu   
2                    Mazda 3 1.5L Luxury  2019.0        490 Triệu   

       assemble_place series  driven kms  num_of_door  num_of_seat  \
0           Nhập khẩu    SUV     65000.0          5.0          5.0   
1  Lắp ráp trong nước    SUV     50000.0          5.0          7.0   
2  Lắp ráp trong nước  Sedan     51000.0          4.0          5.0   

  engine_type transmission  
0        Xăng   Số tự động  
1        Xăng   Số tự động  
2        Xăng   Số tự động

### Drop missing values rows

In [12]:
df.isna().sum()

car_name            0
year              803
price             803
assemble_place    803
series            803
driven kms        803
num_of_door       803
num_of_seat       803
engine_type       807
transmission      806
dtype: int64

- Columns: *year*, *price*, *assemble_place*, *series*, *driven kms*, *num_of_door* each have 803 missing values
- Column *engine_type* has 807 missing values
- Column *transmission* has 806 missing values

In [14]:
# Drop missing values rows
drop_fields = ['year','price','assemble_place','series','driven kms','num_of_door','num_of_seat','engine_type','transmission']
df = df.dropna(subset=drop_fields)

In [15]:
# Check missing values again
df.isna().sum()

car_name          0
year              0
price             0
assemble_place    0
series            0
driven kms        0
num_of_door       0
num_of_seat       0
engine_type       0
transmission      0
dtype: int64

There are no missing values with new DataFrame(26676 rows x 10 cols)

### Drop duplicated rows

In [18]:
df.drop_duplicates(keep="first",inplace= True )

In [19]:
df.index

Index([    0,     1,     2,     3,     4,     5,     6,     7,     8,     9,
       ...
       27421, 27422, 27423, 27424, 27425, 27426, 27427, 27428, 27429, 27430],
      dtype='int64', length=23084)

New DataFrame with 23084 rows x 10 cols

### Data Transformation

Transforming categorical values from Vietnamese to English in columns:
- transmission
- engine_type
- assemble_place

In [23]:
df['transmission'].unique()

array(['Số tự động', 'Số tay'], dtype=object)

In [24]:
df['engine_type'].unique()

array(['Xăng', 'Dầu', 'Hybrid', 'Điện'], dtype=object)

In [25]:
df['assemble_place'].unique()

array(['Nhập khẩu', 'Lắp ráp trong nước'], dtype=object)

In [26]:
df['transmission'] = df['transmission'].replace({'Số tay': 'Manual','Số tự động': 'Automatic'})
df['assemble_place'] = df['assemble_place'].replace({'Nhập khẩu': 'Imported','Lắp ráp trong nước': 'Locally assembled'})
df['engine_type'] = df['engine_type'].replace({'Xăng': 'Petrol','Dầu': 'Diesel','Hybrid': 'Hybrid','điện': 'Electric' })

### Change the data type (dtype)

In [28]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 23084 entries, 0 to 27430
Data columns (total 10 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   car_name        23084 non-null  object 
 1   year            23084 non-null  float64
 2   price           23084 non-null  object 
 3   assemble_place  23084 non-null  object 
 4   series          23084 non-null  object 
 5   driven kms      23084 non-null  float64
 6   num_of_door     23084 non-null  float64
 7   num_of_seat     23084 non-null  float64
 8   engine_type     23084 non-null  object 
 9   transmission    23084 non-null  object 
dtypes: float64(4), object(6)
memory usage: 1.9+ MB


In [29]:
# Change data dtype of columns(year,driven kms,num_of_door,num_of_seat) to int32
df[['year','driven kms','num_of_door','num_of_seat']] = df[['year','driven kms','num_of_door','num_of_seat']].astype(np.int32)

In [30]:
# Change data dtype of columns(transmission,engine_type,assemble_place ) to categorical
col_category = ['transmission','engine_type','assemble_place']
df[col_category] = df[col_category].astype('category')

In [31]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 23084 entries, 0 to 27430
Data columns (total 10 columns):
 #   Column          Non-Null Count  Dtype   
---  ------          --------------  -----   
 0   car_name        23084 non-null  object  
 1   year            23084 non-null  int32   
 2   price           23084 non-null  object  
 3   assemble_place  23084 non-null  category
 4   series          23084 non-null  object  
 5   driven kms      23084 non-null  int32   
 6   num_of_door     23084 non-null  int32   
 7   num_of_seat     23084 non-null  int32   
 8   engine_type     23084 non-null  category
 9   transmission    23084 non-null  category
dtypes: category(3), int32(4), object(3)
memory usage: 1.1+ MB


Why using Categorical Data Type and Int32 Data Type?
- Memory Efficiency
- Optimized processing

### Drop non-validation column ('num_of_door')

In [34]:
df['num_of_door'].unique()

array([ 5,  4,  2,  1,  3,  6, 45, 54, 50,  7])

There are no cars with 45,50,54,6,7 doors.

In [36]:
df = df.drop(['num_of_door'],axis = 1)

### Currency value parsing and conversion for `price` column

In [38]:
def process_price(price):
    try:
        if price.find('Tỷ') != -1:
            ty= price.split('Tỷ')[0]
            trieu = price.split('Tỷ')[1]
            trieu = trieu.split('Triệu')[0]
            return float(ty)*1000000000 + float(trieu)*1000000
        elif price.find('Triệu') != -1:
            trieu = price.split('Triệu')[0]
            trieu = trieu.replace(' ','')
            return float(trieu)*1000000
        else:
            return 0
    except:
        return 0
df['price'] = df['price'].astype(str).apply(process_price)

In [39]:
df.head(3)

car_name  year         price  \
0          LandRover Range Rover HSE 3.0  2015  2.700000e+09   
1   Mitsubishi Outlander 2.0 CVT Premium  2018  6.660000e+08   
2                    Mazda 3 1.5L Luxury  2019  4.900000e+08   

      assemble_place series  driven kms  num_of_seat engine_type transmission  
0           Imported    SUV       65000            5      Petrol    Automatic  
1  Locally assembled    SUV       50000            7      Petrol    Automatic  
2  Locally assembled  Sedan       51000            5      Petrol    Automatic

In [40]:
# Convert currency values from VND to USD
df['price'] = (df['price']/25355).round(0)

In [41]:
df.head(3)

car_name  year     price     assemble_place  \
0          LandRover Range Rover HSE 3.0  2015  106488.0           Imported   
1   Mitsubishi Outlander 2.0 CVT Premium  2018   26267.0  Locally assembled   
2                    Mazda 3 1.5L Luxury  2019   19326.0  Locally assembled   

  series  driven kms  num_of_seat engine_type transmission  
0    SUV       65000            5      Petrol    Automatic  
1    SUV       50000            7      Petrol    Automatic  
2  Sedan       51000            5      Petrol    Automatic

### Save as a new CSV file

In [68]:
df.to_csv('data_carVN_cleaned.csv',index = False)